File: dedup-tracker.mjs
#!/usr/bin/env node
/**
 * dedup-tracker.mjs — Remove duplicate entries from applications.md
 *
 * Groups by normalized company + fuzzy role match.
 * Keeps entry with highest score. If discarded entry had more advanced status,
 * preserves that status. Merges notes.
 *
 * Run: node career-ops/dedup-tracker.mjs [--dry-run]
 */

import { readFileSync, writeFileSync, copyFileSync, existsSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { roleFuzzyMatch } from './role-matcher.mjs';

const CAREER_OPS = dirname(fileURLToPath(import.meta.url));
// Support both layouts: data/applications.md (boilerplate) and applications.md
// (original). CAREER_OPS_TRACKER lets tests point the script at an isolated
// fixture so the real user tracker is never touched.
const APPS_FILE = process.env.CAREER_OPS_TRACKER
  ? process.env.CAREER_OPS_TRACKER
  : existsSync(join(CAREER_OPS, 'data/applications.md'))
    ? join(CAREER_OPS, 'data/applications.md')
    : join(CAREER_OPS, 'applications.md');
const DRY_RUN = process.argv.includes('--dry-run');

// Ensure the target tracker directory exists in both normal and fixture mode.
mkdirSync(dirname(APPS_FILE), { recursive: true });

// Status advancement order (higher = more advanced in pipeline)
// Aplicado > Rechazado because active application > terminal state
const STATUS_RANK = {
  // English canonicals (states.yml labels)
  'skip': 0,
  'discarded': 0,
  'rejected': 1,
  'evaluated': 2,
  'applied': 3,
  'responded': 4,
  'interview': 5,
  'offer': 6,
  // Spanish aliases — kept for backwards compat with existing tracker data
  'no_aplicar': 0,
  'no aplicar': 0,
  'descartado': 0,
  'descartada': 0,
  'rechazado': 1,  // Terminal — below active states
  'rechazada': 1,
  'evaluada': 2,
  'aplicado': 3,
  'respondido': 4,
  'entrevista': 5,
  'oferta': 6,
};

function normalizeCompany(name) {
  return name.toLowerCase()
    .replace(/[()]/g, '')
    .replace(/\s+/g, ' ')
    .replace(/[^a-z0-9 ]/g, '')
    .trim();
}

function normalizeStatus(status) {
  return String(status ?? '')
    .replace(/\*\*/g, '')
    .replace(/\s+\d{4}-\d{2}-\d{2}.*$/, '')
    .trim()
    .toLowerCase();
}

function statusRank(status) {
  return STATUS_RANK[normalizeStatus(status)] || 0;
}

function isAdvancedStatus(status) {
  return statusRank(status) >= STATUS_RANK.applied;
}

function extractReportNum(reportStr) {
  const m = String(reportStr ?? '').match(/\[(\d+)\]/);
  return m ? parseInt(m[1]) : null;
}

function sameReportIdentity(a, b) {
  if (a.num === b.num) return true;
  const reportA = extractReportNum(a.report);
  const reportB = extractReportNum(b.report);
  return reportA !== null && reportA === reportB;
}

function pairKey(a, b) {
  return [a.num, b.num].sort((x, y) => x - y).join(':');
}

const protectedFuzzyPairs = new Set();

function roleMatch(a, b) {
  if (sameReportIdentity(a, b)) return true;
  if (!roleFuzzyMatch(a.role, b.role)) return false;

  // Fuzzy title matches are intentionally conservative once either row has
  // entered the real application pipeline. A user may already have applied to
  // one sibling role, so deleting that row because a higher-scored sibling has
  // similar wording would lose status, report, and notes. Keep both unless the
  // rows point to the exact same report identity.
  if (isAdvancedStatus(a.status) || isAdvancedStatus(b.status)) {
    const key = pairKey(a, b);
    if (!protectedFuzzyPairs.has(key)) {
      protectedFuzzyPairs.add(key);
      console.warn(`⚠️  Keep #${a.num} and #${b.num}: fuzzy role match but advanced status requires exact report identity`);
    }
    return false;
  }

  return true;
}

function parseScore(s) {
  const m = s.replace(/\*\*/g, '').match(/([\d.]+)/);
  return m ? parseFloat(m[1]) : 0;
}

function parseAppLine(line) {
  const parts = line.split('|').map(s => s.trim());
  if (parts.length < 9) return null;
  const num = parseInt(parts[1]);
  if (isNaN(num)) return null;
  return {
    num,
    date: parts[2],
    company: parts[3],
    role: parts[4],
    score: parts[5],
    status: parts[6],
    pdf: parts[7],
    report: parts[8],
    notes: parts[9] || '',
    raw: line,
  };
}

// Read
if (!existsSync(APPS_FILE)) {
  console.log('No applications.md found. Nothing to dedup.');
  process.exit(0);
}
const content = readFileSync(APPS_FILE, 'utf-8');
const lines = content.split('\n');

// Parse all entries
const entries = [];
const entryLineMap = new Map(); // num → line index

for (let i = 0; i < lines.length; i++) {
  if (!lines[i].startsWith('|')) continue;
  const app = parseAppLine(lines[i]);
  if (app && app.num > 0) {
    entries.push(app);
    entryLineMap.set(app.num, i);
  }
}

console.log(`📊 ${entries.length} entries loaded`);

// Group by company+role
const groups = new Map();
for (const entry of entries) {
  const key = normalizeCompany(entry.company);
  if (!groups.has(key)) groups.set(key, []);
  groups.get(key).push(entry);
}

// Find duplicates
let removed = 0;
const linesToRemove = new Set();

for (const [company, companyEntries] of groups) {
  if (companyEntries.length < 2) continue;

  // Within same company, find role matches
  const processed = new Set();
  for (let i = 0; i < companyEntries.length; i++) {
    if (processed.has(i)) continue;
    const cluster = [companyEntries[i]];
    processed.add(i);

    for (let j = i + 1; j < companyEntries.length; j++) {
      if (processed.has(j)) continue;
      if (roleMatch(companyEntries[i], companyEntries[j])) {
        cluster.push(companyEntries[j]);
        processed.add(j);
      }
    }

    if (cluster.length < 2) continue;

    // Keep the one with highest score
    cluster.sort((a, b) => parseScore(b.score) - parseScore(a.score));
    const keeper = cluster[0];

    // Check if any removed entry has more advanced status
    let bestStatusRank = statusRank(keeper.status);
    let bestStatus = keeper.status;
    for (let k = 1; k < cluster.length; k++) {
      const rank = statusRank(cluster[k].status);
      if (rank > bestStatusRank) {
        bestStatusRank = rank;
        bestStatus = cluster[k].status;
      }
    }

    // Update keeper's status if a removed entry had a more advanced one
    if (bestStatus !== keeper.status) {
      const lineIdx = entryLineMap.get(keeper.num);
      if (lineIdx !== undefined) {
        const parts = lines[lineIdx].split('|').map(s => s.trim());
        parts[6] = bestStatus;
        lines[lineIdx] = '| ' + parts.slice(1, -1).join(' | ') + ' |';
        console.log(`  📝 #${keeper.num}: status promoted to "${bestStatus}" (from #${cluster.find(e => e.status === bestStatus)?.num})`);
      }
    }

    // Remove duplicates
    for (let k = 1; k < cluster.length; k++) {
      const dup = cluster[k];
      const lineIdx = entryLineMap.get(dup.num);
      if (lineIdx !== undefined) {
        linesToRemove.add(lineIdx);
        removed++;
        console.log(`🗑️  Remove #${dup.num} (${dup.company} — ${dup.role}, ${dup.score}) → kept #${keeper.num} (${keeper.score})`);
      }
    }
  }
}

// Remove lines (in reverse order to preserve indices)
const sortedRemoveIndices = [...linesToRemove].sort((a, b) => b - a);
for (const idx of sortedRemoveIndices) {
  lines.splice(idx, 1);
}

console.log(`\n📊 ${removed} duplicates removed`);

if (!DRY_RUN && removed > 0) {
  copyFileSync(APPS_FILE, APPS_FILE + '.bak');
  writeFileSync(APPS_FILE, lines.join('\n'));
  console.log('✅ Written to applications.md (backup: applications.md.bak)');
} else if (DRY_RUN) {
  console.log('(dry-run — no changes written)');
} else {
  console.log('✅ No duplicates found');
}

File: scripts/python/tracker/dedup_tracker.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import re
import shutil
from dataclasses import dataclass
from pathlib import Path

from scripts.python.tracker.parse import parse_tracker_row, resolve_columns
from scripts.python.tracker.utils import rebuild_row, resolve_tracker_path


STATUS_RANK = {
    "skip": 0,
    "discarded": 0,
    "rejected": 1,
    "evaluated": 2,
    "applied": 3,
    "responded": 4,
    "interview": 5,
    "offer": 6,
    "no_aplicar": 0,
    "no aplicar": 0,
    "descartado": 0,
    "descartada": 0,
    "rechazado": 1,
    "rechazada": 1,
    "evaluada": 2,
    "aplicado": 3,
    "respondido": 4,
    "entrevista": 5,
    "oferta": 6,
}


@dataclass(frozen=True)
class DedupSummary:
    removed: int
    promoted: int


def normalize_company(name: str) -> str:
    return re.sub(r"[^a-z0-9 ]", "", str(name or "").lower().replace("(", "").replace(")", "")).strip()


def normalize_status(status: str) -> str:
    return re.sub(r"\s+\d{4}-\d{2}-\d{2}.*$", "", str(status or "").replace("**", "")).strip().lower()


def status_rank(status: str) -> int:
    return STATUS_RANK.get(normalize_status(status), 0)


def is_advanced_status(status: str) -> bool:
    return status_rank(status) >= STATUS_RANK["applied"]


def extract_report_num(report: str) -> int | None:
    match = re.search(r"\[(\d+)\]", str(report or ""))
    return int(match.group(1)) if match else None


def same_report_identity(a: dict[str, object], b: dict[str, object]) -> bool:
    if a["num"] == b["num"]:
        return True
    report_a = extract_report_num(str(a.get("report", "")))
    report_b = extract_report_num(str(b.get("report", "")))
    return report_a is not None and report_a == report_b


def normalize_role(role: str) -> str:
    return re.sub(r"\s+", " ", str(role or "")).strip().lower()


def role_match(a: dict[str, object], b: dict[str, object]) -> bool:
    if same_report_identity(a, b):
        return True
    if normalize_role(str(a.get("role", ""))) != normalize_role(str(b.get("role", ""))):
        return False
    return not (is_advanced_status(str(a.get("status", ""))) or is_advanced_status(str(b.get("status", ""))))


def parse_score(value: str) -> float:
    match = re.search(r"([\d.]+)", str(value or "").replace("**", ""))
    return float(match.group(1)) if match else 0.0


def within_blind_window(a: str, b: str, days: int = 90) -> bool:
    from datetime import date

    try:
        da = date.fromisoformat(str(a))
        db = date.fromisoformat(str(b))
    except ValueError:
        return False
    return abs((da - db).days) <= days


def dedup_tracker(tracker_path: str | Path | None = None, *, dry_run: bool = False, backup: bool = True) -> DedupSummary:
    apps_file = Path(tracker_path) if tracker_path else resolve_tracker_path()
    if not apps_file.exists():
        return DedupSummary(0, 0)
    lines = apps_file.read_text(encoding="utf-8").split("\n")
    colmap = resolve_columns(lines)
    entries: list[dict[str, object]] = []
    for idx, line in enumerate(lines):
        row = parse_tracker_row(line, colmap)
        if row and row.num > 0:
            entries.append({**row.__dict__, "line_idx": idx})

    blind_prefix = "\0blind-via:"
    groups: dict[str, list[dict[str, object]]] = {}
    for entry in entries:
        key = blind_prefix + normalize_company(str(entry.get("via", ""))) if str(entry.get("company", "")).strip() == "?" else normalize_company(str(entry.get("company", "")))
        groups.setdefault(key, []).append(entry)

    removed = promoted = 0
    lines_to_remove: set[int] = set()
    for key, group in groups.items():
        if len(group) < 2:
            continue
        is_blind = key.startswith(blind_prefix)
        processed: set[int] = set()
        for i, base in enumerate(group):
            if i in processed:
                continue
            cluster = [base]
            processed.add(i)
            for j in range(i + 1, len(group)):
                candidate = group[j]
                if j in processed:
                    continue
                if role_match(base, candidate) and (not is_blind or within_blind_window(str(base["date"]), str(candidate["date"]))):
                    cluster.append(candidate)
                    processed.add(j)
            if len(cluster) < 2:
                continue
            cluster.sort(key=lambda row: parse_score(str(row.get("score", ""))), reverse=True)
            keeper = cluster[0]
            best_status = max(cluster, key=lambda row: status_rank(str(row.get("status", ""))))["status"]
            if best_status != keeper["status"]:
                parts = [part.strip() for part in lines[int(keeper["line_idx"])].split("|")]
                parts[colmap["status"]] = str(best_status)
                lines[int(keeper["line_idx"])] = rebuild_row(parts)
                promoted += 1
            for duplicate in cluster[1:]:
                lines_to_remove.add(int(duplicate["line_idx"]))
                removed += 1

    for idx in sorted(lines_to_remove, reverse=True):
        lines.pop(idx)
    if removed and not dry_run:
        if backup:
            shutil.copyfile(apps_file, Path(str(apps_file) + ".bak"))
        apps_file.write_text("\n".join(lines), encoding="utf-8")
    return DedupSummary(removed, promoted)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Remove duplicate applications.md entries.")
    parser.add_argument("--tracker")
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--no-backup", action="store_true")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    summary = dedup_tracker(args.tracker, dry_run=args.dry_run, backup=not args.no_backup)
    print(f"{summary.removed} duplicates removed")
    if args.dry_run:
        print("(dry-run — no changes written)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
